### PR TITLE
Batch requestWithdrawals

### PR DESCRIPTION
**Pull request by to-sh-1**
https://github.com/shapeshift/foxy/issues/21

- Added sendWithdrawalRequests function to send batched withdrawal requests based on cycle times
- Changed block time as well as reset between tests to speed up tests
- Due to new canBatchTransactions function, added a test function to mine to appropriate time to withdraw.

### CODE REVIEW

File: src/contracts/Staking.sol [hunk 1]
```diff
@@ -100,11 +100,21 @@ interface ITokePool {
     function balanceOf(address owner) external view returns (uint256);
 }
 
+interface ITokeManager {
```
Comment (0xean) [inline src/contracts/Staking.sol:21]: I wold probably stick this in its own file under src/interfaces/ITokeManager.sol and import it.

File: src/contracts/Staking.sol [hunk 2]
```diff
@@ -116,14 +126,27 @@ contract Staking is Ownable {
     }
     Epoch public epoch;
 
+    struct Claim {
+        uint256 amount;
+        uint256 gons;
+        uint256 expiry;
+        bool lock; // prevents malicious delays
+    }
+    mapping(address => Claim) public warmupInfo;
```
Comment (0xean) [inline src/contracts/Staking.sol]: same minor nit picks on the naming convention, i think we should do a find replace and make all of these `warmUpInfo` `coolDownInfo` to stick with strict camelCase.

File: src/contracts/Staking.sol [hunk 3]
```diff
@@ -116,14 +126,27 @@ contract Staking is Ownable {
     }
     Epoch public epoch;
 
+    struct Claim {
```
Comment (0xean) [inline src/contracts/Staking.sol:49]: We _could_ probably tightly pack this struct, maybe something we can save as a TODO / GH issue to look at in the end for gas optimization.

File: src/contracts/Staking.sol [hunk 4]
```diff
@@ -207,6 +223,35 @@ contract Staking is Ownable {
         return tokePoolContract.balanceOf(address(this)); // TODO: verify pending withdraws are a part of this
     }
 
+    /**
+        @notice checks TOKE's cycleTime is withink duration to batch the transactions
+        @return bool
+     */
+    function canBatchTransactions() internal view returns (bool) {
+        ITokeManager iTokeManager = ITokeManager(tokeManager);
+        uint256 offset = 50; // amount of blocks before the next cycle to batch the withdrawal requests
```
Comment (0xean) [inline src/contracts/Staking.sol:152]: How do we update this when their cycles change?

**Review by 0xean — APPROVED** [review]
👍  - generally this looks sane to me. A few things to think about.  I believe there is N states we need to worry about:

1.  We are in current cycle with Tokemak, we have not requested our withdraw
2.  We are in current cycle with Tokemak, we have requested our withdrawal
3.  We are in the "next" cycle with Tokemake, and we missed our withdrawal from the last cycle for some error that occured (bot was broken, no users called the contract, etc).

I am not sure we have a test covering state 3, but we should and see what we can do to handle that appropriately. 

Also, I am not clear on what state we end up in if we miss our withdraw from tokemak, maybe you already have this covered, just something to think about.  

One last consideration:
Stakers are already implicitly trusting the DAO (owner) to provide the needed rewards to keep this whole cycle going for rewards. It may make sense to build in a few state functions that the owner can have privileged access to.  I am usually against this, but in our case it may make sense if the owner could do the following:

1. Pause / unpause deposits
2. Pause / unpause withdrawals
3. Have a privileged override to ignore the cycle logic and request the withdraw of FOX from tokemak and withdraw from tokemak
4. pull the toke rewards (I think this is already will be publicly available for anyone to call). 

If we messed this all up, alteast the DAO could then pause functionality, get the FOX out of tokemak, and re-enable withdrawals 

Thoughts?

File: src/contracts/Staking.sol (hunk 4 above)
Comment (to-sh-1) [reply src/contracts/Staking.sol:152]: the offset is just the amount of blocks before their next cycle.  We get the cycle start time and duration directly from the toke manager contract.  Is this what you were asking?

File: src/contracts/Staking.sol (hunk 1 above)
Comment (to-sh-1) [reply src/contracts/Staking.sol:21]: Gonna add these in the claim PR because there's some changes around them

**Comment by to-sh-1** [comment]
> 👍 - generally this looks sane to me. A few things to think about. I believe there is N states we need to worry about:
> 
> 1. We are in current cycle with Tokemak, we have not requested our withdraw
> 2. We are in current cycle with Tokemak, we have requested our withdrawal
> 3. We are in the "next" cycle with Tokemake, and we missed our withdrawal from the last cycle for some error that occured (bot was broken, no users called the contract, etc).
> 
> I am not sure we have a test covering state 3, but we should and see what we can do to handle that appropriately.
> 
> Also, I am not clear on what state we end up in if we miss our withdraw from tokemak, maybe you already have this covered, just something to think about.
> 
> One last consideration: Stakers are already implicitly trusting the DAO (owner) to provide the needed rewards to keep this whole cycle going for rewards. It may make sense to build in a few state functions that the owner can have privileged access to. I am usually against this, but in our case it may make sense if the owner could do the following:
> 
> 1. Pause / unpause deposits
> 2. Pause / unpause withdrawals
> 3. Have a privileged override to ignore the cycle logic and request the withdraw of FOX from tokemak and withdraw from tokemak
> 4. pull the toke rewards (I think this is already will be publicly available for anyone to call).
> 
> If we messed this all up, alteast the DAO could then pause functionality, get the FOX out of tokemak, and re-enable withdrawals
> 
> Thoughts?

If we miss the withdraws on that cycle, it will just happen on the next cycle.  We are saving the last index that we requested a withdraw so there shouldn't be an ability to do it twice in a cycle.  

I think functions to override the logic would be great.  If things go wrong we could at least let the owner of the contract get the funds out.  Also I plan on implementing some code coverage and getting it higher than it is.  Please let me know of any situations you see that haven't been tested yet.  I haven't done a deep dive in regards to testing, but I plan to this week

File: src/contracts/Staking.sol (hunk 4 above)
Comment (0xean) [reply src/contracts/Staking.sol:152]: yea, I am asking when they change their cycles, I assume we need to update this offset, right?

Comment (to-sh-1) [reply src/contracts/Staking.sol:152]: I dm'd you about this, but I'm going to go ahead and merge this as it's the first of the 3 PR's that are available.  If we need to make changes we can

**Pull request merged**